src/lib.rs

```diff
@@ -658,6 +698,91 @@ impl<'a> std::io::Write for GzipLayerWriter<'a> {
     }
 }
 
+#[cfg(feature = "zstd")]
+impl<'a> ZstdLayerWriter<'a> {
+    /// Create a writer for a gzip compressed layer blob.
+    fn new(ocidir: &'a Dir, c: Option<i32>) -> Result<Self> {
+        let bw = BlobWriter::new(ocidir)?;
+        let encoder = zstd::Encoder::new(bw, c.unwrap_or(0))?;
+        Ok(Self(Sha256Writer::new(encoder)))
+    }
+
+    /// Consume this writer, flushing buffered data and put the blob in place.
+    pub fn complete(self) -> Result<Layer> {
+        let (digest, enc) = self.0.finish();
+        let blob = enc.finish()?.complete()?;
+        let uncompressed_sha256 = Sha256Digest::from_str(&digest).unwrap();
+        Ok(Layer {
+            blob,
+            uncompressed_sha256,
+            media_type: MediaType::ImageLayerZstd,
+        })
+    }
+}
+
+#[cfg(feature = "zstdmt")]
+impl<'a> ZstdLayerWriter<'a> {
+    /// Create a writer for a zstd compressed layer blob, with multithreaded compression enabled.
+    ///
+    /// The `n_workers` parameter specifies the number of threads to use for compression, per
+    /// [Encoder::multithread]]
+    fn multithread(ocidir: &'a Dir, c: Option<i32>, n_workers: u32) -> Result<Self> {
+        let bw = BlobWriter::new(ocidir)?;
+        let mut encoder = zstd::Encoder::new(bw, c.unwrap_or(0))?;
+        encoder.multithread(n_workers)?;
+        Ok(Self(Sha256Writer::new(encoder)))
+    }
+}
+
+#[cfg(feature = "zstd")]
+impl<'a> std::io::Write for ZstdLayerWriter<'a> {
+    fn write(&mut self, data: &[u8]) -> std::io::Result<usize> {
+        self.0.write(data)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.flush()
+    }
+}
+
+#[cfg(feature = "zstd")]
+/// Wraps a writer and calculates the sha256 digest of data written to the inner writer
+struct Sha256Writer<W> {
```

Comment:
Hmm this could be factored out and reused for gzip too right?

Reply:
yes, done. didn't do it for blobwriter since that also determines the size (and I thought adding a SizeWriter that keeps track of data size, and making blobwriter have a SizeWriter<Sha256Writer> or similar was too much :) )